### PR TITLE
Expanded documentation for the ApplicationBuilder

### PR DESCRIPTION
**Pull request by csherratt**
I think I am done documenting all the functions in this module. Please tell me if there is anything that you think should be expanded on.

### CODE REVIEW

File: src/app.rs [hunk 1]
```diff
@@ -18,9 +18,12 @@ use error::{Error, Result};
 use state::{State, StateMachine};
 use timing::{Stopwatch, Time};
 
-/// User-friendly facade for building games. Manages main loop.
+/// An Application is the root objects of the game engine. It binds the OS
```
Comment (Rhuagh) [inline src/app.rs]: s/objects/object/

File: src/app.rs [hunk 2]
```diff
@@ -207,7 +207,9 @@ impl<'a, 'b> Drop for Application<'a, 'b> {
     }
 }
 
-/// Helper builder for Applications.
+/// `ApplicationBuilder` is an interface that allows for creation of a [`Application`](struct.Application.html)
```
Comment (Rhuagh) [inline src/app.rs]: s/a/an/

File: src/app.rs [hunk 3]
```diff
@@ -258,31 +260,28 @@ impl<'a, 'b, T: State + 'a> ApplicationBuilder<'a, 'b, T> {
...
-    ///         .build()
-    ///         .expect("Failed to create Application");
-    ///
-    ///     // the game instance can now be run, this exits only when the game is done
-    ///     game.run();
-    /// }
+    /// // initialize the builder, the `ApplicationBuilder` object
+    /// // follows the use pattern of most builder objects found
+    /// // in the rust ecosystem. Each function modifies the object
+    /// // returning a new object that with the modified configuration.
```
Comment (Rhuagh) [inline src/app.rs]: s/that with/with/

File: src/app.rs [hunk 4]
```diff
@@ -426,57 +421,278 @@ impl<'a, 'b, T: State + 'a> ApplicationBuilder<'a, 'b, T> {
...
+    ///
+    /// struct NopSystem;
+    /// impl<'a> System<'a> for NopSystem {
+    ///     type SystemData = ();
+    ///     fn run(&mut self, (): Self::SystemData) {}
+    /// }
+    ///
+    /// // Three systems are added in this example. The "tabby cat" & "tom cat"
+    /// // systems will both run in parallel. Only after both cat systems have
+    /// // run is the "doggo" system permitted to run after them.
```
Comment (Rhuagh) [inline src/app.rs]: s/after//

File: src/app.rs [hunk 5]
```diff
@@ -426,57 +421,278 @@ impl<'a, 'b, T: State + 'a> ApplicationBuilder<'a, 'b, T> {
...
 
-    /// Adds a given system `sys`, assigns it the string identifier `name`,
-    /// and marks it dependent on systems `dep`.
-    /// Note: all dependencies should be added before you add depending system
-    /// If you want to register systems which can not be specified as dependencies,
-    /// you can use "" as their name, which will not panic (using another name twice will).
-    pub fn with<S>(mut self, sys: S, name: &str, dep: &[&str]) -> Self
+    /// Adds a given system to the game loop.
+    ///
+    /// __Note:__ all dependencies should be added before you add depending system.
```
Comment (Rhuagh) [inline src/app.rs]: s/should/must/
s/depending/the/

File: src/app.rs [hunk 6]
```diff
@@ -426,57 +421,278 @@ impl<'a, 'b, T: State + 'a> ApplicationBuilder<'a, 'b, T> {
...
+    /// This function returns ApplicationBuilder after it has modified it.
+    ///
+    /// # Type Parameters
+    ///
+    /// - `S`: A type that implements the `System` trait.
+    ///
+    /// # Panics
+    /// 
+    /// If two system are added that share an identical name, this function will panic.
+    /// Empty names are permitted, and this function will not panic of more then two are added.
```
Comment (Rhuagh) [inline src/app.rs]: s/panic of/panic if/

File: src/app.rs [hunk 7]
```diff
@@ -426,57 +421,278 @@ impl<'a, 'b, T: State + 'a> ApplicationBuilder<'a, 'b, T> {
...
+        self.disp_builder = self.disp_builder.add(system, name, dependencies);
         self
     }
 
-    /// Adds a given thread-local system `sys` to the Application.
+    /// Add a given thread-local system to the game loop.
+    ///
+    /// A thread-local system is one that _must_ be run on the main thread of 
+    /// the game. This is normally used to work around thread unsafe APIs provided
+    /// by some OS vendors. Example being OpenGL and many UI polling requires
```
Comment (Rhuagh) [inline src/app.rs]: This sentence feels a bit weird, not sure how to fix it though.

File: src/app.rs [hunk 8]
```diff
@@ -426,57 +421,278 @@ impl<'a, 'b, T: State + 'a> ApplicationBuilder<'a, 'b, T> {
...
+    /// - Failure to create a window
+    /// - Failure to initialize the lower level graphics context (OpenGL, DX)
+    /// - Failure to initialize the thread pool
+    /// - Failure to compile shaders for the target platform
+    /// - linking error between shader variables and framework
+    ///
+    /// # Examples
+    ///
+    /// See the `renderable` example supplied in the examples folder for a fully
+    /// working example that illustrates how to configuration of the renderer.
```
Comment (Rhuagh) [inline src/app.rs]: /configuration of/configure/

File: src/app.rs [hunk 9]
```diff
@@ -426,57 +421,278 @@ impl<'a, 'b, T: State + 'a> ApplicationBuilder<'a, 'b, T> {
...
+    /// Register an asset store with the loader logic of the Application.
+    ///
+    /// If the asset store exists, that shares a name with the new store the net
+    /// effect will be a replacement of the older store with the new one.
+    /// No warning or panic will result from this action.
+    ///
+    /// # Parameters
+    ///
+    /// - `name`: A unique name or key to identify the asset storage location. `name`
+    ///           is used later during asset loading to determine the registered asset store.
```
Comment (Rhuagh) [inline src/app.rs]: to specify where the asset should be loaded from.

File: src/app.rs [hunk 10]
```diff
@@ -489,7 +705,32 @@ impl<'a, 'b, T: State + 'a> ApplicationBuilder<'a, 'b, T> {
         self
     }
 
-    /// Register new context within the loader
+    /// Register a new asset type with the Application. All required components
+    /// related to the storage of this asset type will be registered. Since
+    /// Amethyst uses AssetFutures to allow for async content loading, Amethyst
+    /// needs to have a system that translates AssetFutures into Components as
+    /// they resolve. Amethyst includes a system to avoid the user having to
```
Comment (Rhuagh) [inline src/app.rs]: Technically, I think it's included in Specs and not directly in Amethyst?

Comment (csherratt) [reply src/app.rs]: Ok, ill clarify.

**Comment by csherratt** [comment]
Updated based on @Rhuagh's comments.

**Pull request merged**